File: Sources/SajiloApp/Models/Rashifal.swift
import Foundation

/// One sign's reading for the day.
struct Rashifal: Codable, Equatable, Sendable, Identifiable {
    let sign: RashiSign
    /// The astrologer's words, carried verbatim. Never trimmed, summarised, or
    /// reflowed — it is someone's writing, and Sajilo shows it as published.
    let prediction: String

    var id: String { sign.rawValue }
}

/// The twelve rashi, in their canonical order.
///
/// These are the same twelve divisions as the Bikram Sambat solar months —
/// Baishakh is Mesh, Jestha is Vrish, and so on — because a BS month is the
/// span the sun spends in one rashi.
///
/// That correspondence is deliberately *not* used to pick a reader's sign for
/// them. In Nepali practice your rashi is normally the **moon** sign from your
/// birth chart, or the one a jyotish assigned from the first syllable of your
/// name; it is not your birth month. Deriving it from a birth date would give
/// the sun sign and quietly hand most people the wrong reading, so the sign is
/// always chosen by hand.
enum RashiSign: String, Codable, Equatable, Sendable, CaseIterable, Identifiable {
    case mesh, vrish, mithun, karkat, simha, kanya
    case tula, vrishchik, dhanu, makar, kumbha, meen

    var id: String { rawValue }

    /// As Hamro Patro heads each section.
    var nepaliName: String {
        switch self {
        case .mesh: "मेष"
        case .vrish: "वृष"
        case .mithun: "मिथुन"
        case .karkat: "कर्कट"
        case .simha: "सिंह"
        case .kanya: "कन्या"
        case .tula: "तुला"
        case .vrishchik: "वृश्चिक"
        case .dhanu: "धनु"
        case .makar: "मकर"
        case .kumbha: "कुम्भ"
        case .meen: "मीन"
        }
    }

    /// Romanised Nepali rather than the Western equivalent: someone who knows
    /// they are Mesh does not necessarily think of themselves as Aries.
    var romanName: String {
        switch self {
        case .mesh: "Mesh"
        case .vrish: "Vrish"
        case .mithun: "Mithun"
        case .karkat: "Karkat"
        case .simha: "Simha"
        case .kanya: "Kanya"
        case .tula: "Tula"
        case .vrishchik: "Vrishchik"
        case .dhanu: "Dhanu"
        case .makar: "Makar"
        case .kumbha: "Kumbha"
        case .meen: "Meen"
        }
    }

    /// The Western name, for readers who only know that one.
    var westernName: String {
        switch self {
        case .mesh: "Aries"
        case .vrish: "Taurus"
        case .mithun: "Gemini"
        case .karkat: "Cancer"
        case .simha: "Leo"
        case .kanya: "Virgo"
        case .tula: "Libra"
        case .vrishchik: "Scorpio"
        case .dhanu: "Sagittarius"
        case .makar: "Capricorn"
        case .kumbha: "Aquarius"
        case .meen: "Pisces"
        }
    }

    /// The zodiac sign as a character rather than an image. SF Symbols has no
    /// zodiac glyphs at all, and the plain code points (U+2648…) default to
    /// *emoji* presentation — coloured tiles that ignore `foregroundStyle` and
    /// fight the palette. The trailing U+FE0E forces text presentation, so the
    /// glyph takes the brass like any other type.
    var glyph: String {
        let scalar: Unicode.Scalar = switch self {
        case .mesh: "\u{2648}"
        case .vrish: "\u{2649}"
        case .mithun: "\u{264A}"
        case .karkat: "\u{264B}"
        case .simha: "\u{264C}"
        case .kanya: "\u{264D}"
        case .tula: "\u{264E}"
        case .vrishchik: "\u{264F}"
        case .dhanu: "\u{2650}"
        case .makar: "\u{2651}"
        case .kumbha: "\u{2652}"
        case .meen: "\u{2653}"
        }
        return "\(scalar)\u{FE0E}"
    }

    /// The नामाक्षर — the syllables a name beginning with places you in this
    /// rashi. This is how most Nepalis actually know their sign, so it is what
    /// the picker is built around.
    ///
    /// Bundled rather than fetched: these are traditional and identical across
    /// every jyotish source, so they never go stale, and the picker works
    /// before any network call has returned.
    var namingSyllables: [String] {
        switch self {
        case .mesh: ["चु", "चे", "चो", "ला", "लि", "लु", "ले", "लो", "अ"]
        case .vrish: ["इ", "उ", "ए", "ओ", "वा", "वि", "वु", "वे", "वो"]
        case .mithun: ["का", "कि", "कु", "घ", "ङ", "छ", "के", "को", "हा"]
        case .karkat: ["हि", "हु", "हे", "हो", "डा", "डि", "डु", "डे", "डो"]
        case .simha: ["मा", "मि", "मु", "मे", "मो", "टा", "टि", "टु", "टे"]
        case .kanya: ["टो", "पा", "पि", "पु", "ष", "ण", "ठ", "पे", "पो"]
        case .tula: ["रा", "रि", "रु", "रे", "रो", "ता", "ति", "तु", "ते"]
        case .vrishchik: ["तो", "ना", "नि", "नु", "ने", "नो", "या", "यि", "यु"]
        case .dhanu: ["ये", "यो", "भा", "भि", "भु", "धा", "फा", "ढा", "भे"]
        case .makar: ["भो", "जा", "जि", "जु", "जे", "जो", "ख", "खि", "खु", "खे", "खो", "गा", "गि"]
        case .kumbha: ["गु", "गे", "गो", "सा", "सि", "सु", "से", "सो", "दा"]
        case .meen: ["दि", "दु", "थ", "झ", "ञ", "दे", "दो", "चा", "चि"]
        }
    }
}

struct RashifalSnapshot: Codable, Equatable, Sendable {
    let readings: [Rashifal]
    /// The Bikram Sambat day the source published this for, so a reading left
    /// over from yesterday can say so instead of passing as today's.
    let publishedOn: NepaliDate?
    let fetchedAt: Date

    func reading(for sign: RashiSign) -> Rashifal? {
        readings.first { $0.sign == sign }
    }
}

File: crates/sajilo-api/src/rashifal.rs
//! Daily rashifal readings. Ported from `Rashifal.swift`.

use crate::load_state::Freshness;

dto_enum! {
    /// The twelve rashi, in their canonical order.
    ///
    /// These are the same twelve divisions as the Bikram Sambat solar months,
    /// because a BS month is the span the sun spends in one rashi. That
    /// correspondence is deliberately *not* used to pick a reader's sign for
    /// them: in Nepali practice your rashi is normally the **moon** sign from
    /// your birth chart, or the one a jyotish assigned from the first syllable
    /// of your name. Deriving it from a birth date would give the sun sign and
    /// quietly hand most people the wrong reading, so the sign is always
    /// chosen by hand.
    pub enum RashiSign {
        Mesh,
        Vrish,
        Mithun,
        Karkat,
        Simha,
        Kanya,
        Tula,
        Vrishchik,
        Dhanu,
        Makar,
        Kumbha,
        Meen,
    }
}

dto! {
    /// One sign's reading for the day.
    pub struct Rashifal {
        pub sign: RashiSign,
        /// The astrologer's words, carried verbatim. Never trimmed,
        /// summarised, or reflowed — it is someone's writing, and Sajilo shows
        /// it as published.
        pub prediction: String,
    }

    pub struct RashifalSnapshot {
        pub readings: Vec<Rashifal>,
        pub freshness: Freshness,
    }
}

impl RashiSign {
    pub const ALL: [Self; 12] = [
        Self::Mesh,
        Self::Vrish,
        Self::Mithun,
        Self::Karkat,
        Self::Simha,
        Self::Kanya,
        Self::Tula,
        Self::Vrishchik,
        Self::Dhanu,
        Self::Makar,
        Self::Kumbha,
        Self::Meen,
    ];

    /// As Hamro Patro heads each section.
    pub fn nepali_name(self) -> &'static str {
        match self {
            Self::Mesh => "मेष",
            Self::Vrish => "वृष",
            Self::Mithun => "मिथुन",
            Self::Karkat => "कर्कट",
            Self::Simha => "सिंह",
            Self::Kanya => "कन्या",
            Self::Tula => "तुला",
            Self::Vrishchik => "वृश्चिक",
            Self::Dhanu => "धनु",
            Self::Makar => "मकर",
            Self::Kumbha => "कुम्भ",
            Self::Meen => "मीन",
        }
    }

    /// Romanised Nepali rather than the Western equivalent: someone who knows
    /// they are Mesh does not necessarily think of themselves as Aries.
    pub fn display_name(self) -> &'static str {
        match self {
            Self::Mesh => "Mesh",
            Self::Vrish => "Vrish",
            Self::Mithun => "Mithun",
            Self::Karkat => "Karkat",
            Self::Simha => "Simha",
            Self::Kanya => "Kanya",
            Self::Tula => "Tula",
            Self::Vrishchik => "Vrishchik",
            Self::Dhanu => "Dhanu",
            Self::Makar => "Makar",
            Self::Kumbha => "Kumbha",
            Self::Meen => "Meen",
        }
    }
}

impl RashifalSnapshot {
    pub fn reading(&self, sign: RashiSign) -> Option<&Rashifal> {
        self.readings.iter().find(|reading| reading.sign == sign)
    }
}
